synapse/handlers/e2e_room_keys.py

```diff
@@ -58,7 +61,9 @@ async def get_room_keys(
         version: str,
```

Comment:
[Alarm bells](https://youtu.be/HcFd5j1cios?t=24).

```
synapse=> \d e2e_room_keys_versions
         Table "public.e2e_room_keys_versions"
  Column   |   Type   | Collation | Nullable | Default 
-----------+----------+-----------+----------+---------
 user_id   | text     |           | not null | 
 version   | bigint   |           | not null | 
 algorithm | text     |           | not null | 
 auth_data | text     |           | not null | 
 deleted   | smallint |           | not null | 0
 etag      | bigint   |           |          | 
Indexes:
    "e2e_room_keys_versions_idx" UNIQUE, btree (user_id, version)
```

version is a bigint in the db, so I'm surprised to see a string here. What's going on?

Reply:
Oh. It looks like version comes from a query parameter.

https://github.com/matrix-org/synapse/blob/d1f1b46c2cfe612790aebdd54765953951c94e45/synapse/rest/client/room_keys.py#L197

Can we change this to `parse_integer` at the rest layer, then pass an integer to and from the handler and the DB layer?

Reply:
Normally I'd agree: stringly-typed programming is a code smell.

However the matrix spec says that `version` is an opaque string identifying a backup: https://spec.matrix.org/v1.1/client-server-api/#post_matrixclientv3room_keysversion
`version`s are only issued when a backup is uploaded, by the homeserver. As such, the correct response to a non-numeric version is a 404 (no such backup) rather than the 400 `parse_integer` would raise.

Because the spec says `version`s are opaque identifiers, I'd consider versions-being-integers to be an implementation detail. Only the code responsible for issuing them, `EndToEndRoomKeyStore`, has to know about it. Keeping that detail contained within the data store does have a benefit: if we were to decide that we wanted to use, say, sha256 hashes instead of incrementing integers, we'd only need to update the data store and not the handler and rest layers.

Reply:
It might make sense to include a comment somewhere in the store method saying why we convert it to an int there?

Reply:
Having said that, the handling for non-numeric versions is not pretty code.

Reply:
> It might make sense to include a comment somewhere in the store method saying why we convert it to an int there?

Good call. I think it ought to be a class-level docstring describing the design choice, since it applies to every method.

Reply:
That's a surprise---thanks for digging it up!